Source/Scene/Expression.js

```diff
@@ -0,0 +1,1904 @@
+
+        // customize jsep operators
+        jsep.addBinaryOp('=~', 0);
+        jsep.addBinaryOp('!~', 0);
+
+        var ast;
+        try {
+            ast = jsep(expression);
+        } catch (e) {
+            //>>includeStart('debug', pragmas.debug);
+            throw new DeveloperError(e);
+            //>>includeEnd('debug');
+        }
+
+        this._runtimeAst = createRuntimeAst(this, ast);
+    }
+
+    defineProperties(Expression.prototype, {
+        /**
+         * Gets the expression defined in the 3D Tiles Styling language.
+         *
+         * @memberof Expression.prototype
+         *
+         * @type {String}
+         * @readonly
+         *
+         * @default undefined
+         */
+        expression : {
+            get : function() {
+                return this._expression;
+            }
+        }
+    });
+
+    /**
+     * Evaluates the result of an expression, optionally using the provided feature's properties. If the result of
+     * the expression in the
+     * {@link https://github.com/AnalyticalGraphicsInc/3d-tiles/tree/master/Styling|3D Tiles Styling language}
+     * is of type <code>Boolean</code>, <code>Number</code>, or <code>String</code>, the corresponding JavaScript
+     * primitive type will be returned. If the result is a <code>RegExp</code>, a Javascript <code>RegExp</code>
+     * object will be returned. If the result is a <code>Cartesian2</code>, <code>Cartesian3</code>, or <code>Cartesian4</code>,
+     * a {@link Cartesian2}, {@link Cartesian3}, or {@link Cartesian4} object will be returned.
+     *
+     * @param {FrameState} frameState The frame state.
+     * @param {Cesium3DTileFeature} feature The feature whose properties may be used as variables in the expression.
+     * @returns {Boolean|Number|String|RegExp|Cartesian2|Cartesian3|Cartesian4} The result of evaluating the expression.
+     */
+    Expression.prototype.evaluate = function(frameState, feature) {
+        ScratchStorage.reset();
+        var result = this._runtimeAst.evaluate(frameState, feature);
+        if ((result instanceof Cartesian2) || (result instanceof Cartesian3) || (result instanceof Cartesian4)) {
+            return result.clone();
+        }
+        return result;
+    };
+
+    /**
+     * Evaluates the result of a Color expression, using the values defined by a feature.
+     * <p>
+     * This is equivalent to {@link StyleExpression#evaluate} but avoids allocating memory by accepting a result argument.
+     * </p>
+     * @param {FrameState} frameState The frame state.
+     * @param {Cesium3DTileFeature} feature The feature whose properties may be used as variables in the expression.
+     * @param {Color} [result] The object in which to store the result
+     * @returns {Color} The modified result parameter or a new Color instance if one was not provided.
+     */
+    Expression.prototype.evaluateColor = function(frameState, feature, result) {
```

Comment:
Why not just have `evaluate` take an optional result argument that is valid for all clonable objects?  This is how the Entity API works with Property.getValue.  Wouldn't it be beneficial to use result parameters with Cartesian types for instance?